### ISSUE TITLE
Terraform really needs a way of securely targetting 

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#22468]
Terraform 0.11.14 (but probably all versions)

I have to use -target in some circumstances to apply certain changes in a specific area. 

Why? Because I work on a multi-cloud multi-region platform with hundreds of tenants, spanning hundreds of aws accounts, gcp projects etc. etc., this isn't a trivial terraform installation. Part of such a huge model usually includes some kind of shared-tenancy and shared-ownership, meaning I have to make changes sometimes to terraform resources, which other people can also alter. This is just a fact on the ground, it isnt going to change.

With some use of ignore_changes I can limit the kinds of updates that terraform will think it needs to do if it comes across changes from what it knows (in this model), and sometimes I find I need to use -target applies to get a job done.

I find it bad that -target doesnt have a mode where terraform evaluates that what it is about to do goes outside the scope of the -target, and fails gracefully for further investigation. 

So lets suppose, on a vpc I -target a vpc_flow_logs resource, because my shared-ownership vpc has governance requirements. Terraform needs to be able to refresh the vpc resource and can find that some things have changed - I am only interested in managing the vpc flow logs however, I want to instruct terraform that in no circumstances change the vpc, just the vpc flow logs. There is no way to do this.

The dependency hierarchy of the vpc_flow_logs will be looked up (ok), refreshed (also ok), and worked out what delta would be required for terraform to become authoritative (also ok), but then terraform goes and changes everything it wants, regardless of what i wanted targetting. 

This is also ok for default behaviour, but at the very minimum, there should be a way of telling terraform "no, really dont change anything other than what ive told you, and if you cant do that, or feel uncomfortable about it, then bail out."

When you . have a CI pipeline of hundreds and hundreds of terraform installations, and you need to iterate through them in parallel and get a change done over many different end clients, you need the tools to do the job. As it stands, I have to write code to terraform plan, check that the changes are only the changes affecting the resources i wanted in the original -target anyway, and only proceed to apply the plan on that basis. I'd prefer terraform does the work

I appreciate many wont have this problem because they might operate terraform not at scale, or under complex conditions like this, but often I dont get a choice in some of this setup, and I find it a bit rude (if rude can be used for a tool!?) that terraform can do more than you ask it to on a -target . (and potentially quite dangerous too)

**Comment by hashibot** [comment]
Hello! :wave: 

This issue doesn't seem to be following one of the standard issue templates available from [the "New Issue" page](https://github.com/hashicorp/terraform/issues/new/choose). The information requested in those templates is important for my human friends to fully understand what's going on.

Please feel free to open a new issue using either the Bug Report or Feature Request template, as appropriate, and then the Terraform maintainers will review it. Thanks!

Please note that we use GitHub issues for tracking bugs and enhancements rather than for questions. While we may be able to help with certain simple problems here it's better to use [the community forum](https://discuss.hashicorp.com/c/terraform-core), where there are more people ready to help. The GitHub issues here are generally monitored only by our few core maintainers.

**Comment by teamterraform** [comment]
A way to meet this sort of situation-specific requirement today is:

* Run `terraform plan -out=tfplan` to construct an unconstrained plan and save it to disk.
* Run `terraform show -json tfplan` to produce a machine-readable description of the plan.
* Use some custom logic written in a language of your choice to parse that JSON and ensure that the plan only affects resources you expect; if not, bail out with an error and don't continue.
* If the previous step succeeded, then run `terraform apply tfplan` to apply the plan.

Terraform's workflow is designed around architectures that are constructed from many small Terraform configurations that can be independently applied rather than very large configurations where you must use `-target` to reasonably work with them. However, we implemented this JSON plan format mechanism as a pragmatic way to allow Terraform to be used in situations outside of its core design space, such as the use-case discussed here. We don't have any plans to introduce built-in features aimed at managing very large configurations, but we do encourage those with unusual requirements to consider using Terraform as a component of a broader solution that includes other software to achieve the desired goal.

**Comment by gtmtech** [comment]
@teamterraform  - I know about the desired smaller architectures argument - I've been told this before, and indeed I do that. Although we manage a vastly huge estate, we have many independently apply-able codebases that construct the whole... So apologies if you misunderstood that my terraform apply was very large, not its not (maybe 10-15 resources) - AND YET - it is impossible to make changes that I need to do with terraform alone. The above case is a case in point - my ENTIRE terraform codebase is a VPC, a transit gateway connection, 3 subnets, and a default route table (for the TGW). This is the natural place to add a vpc flowlog. But I cant do it, because of invasive changes terraform wants to reset on the VPC itself. 

The issue(s) normally come because we manage some tag(s) on a resource inferring some kind of semantic meaning, and clients manage other tag(s) on a resource for their own semantic meaning purposes, and there's just no way to do it in terraform regardless of how you divide up your repositories. 

This makes me sad, it's an awesome tool. I guess you will say you are catering for the lions share of use-cases though, so I understand if you're not interested in fixing this one.
